django/template/base.py

```diff
@@ -86,7 +86,8 @@
 # Match BLOCK_TAG_*, VARIABLE_TAG_*, and COMMENT_TAG_* tags and capture the
 # entire tag, including start/end delimiters. Using re.compile() is faster
 # than instantiating SimpleLazyObject with _lazy_re_compile().
-tag_re = re.compile(r"({%.*?%}|{{.*?}}|{#.*?#})")
+tag_re = re.compile(r"({%.*?%}|{{.*?}}|{#.*?#})", re.DOTALL)
```

Comment:
It's worth noting that this change also introduces multi-line variable tokens and multi-line comment tokens, as well as multiline block tokens.

For multi-line variables to work we would also need to update the filter-expression regular expression. Even if we limited multi-line filter expressions to strictly those defined inside variable tokens, this would introduce a lot of additional complexity, and further potential breaking changes without any clear benefit. (There is a definite need for multi-line template tags, but I think a multi-line filter-expression would be less useful). As such, I think it would be better if the lexer just didn't accept them, rather than adding complexity to the parser, which would need to handle such tokens gracefully.

With regards multi-line comments, allowing new lines between `{# ... #}` tags would certainly be a nice syntax to allow. It would be an additional breaking change though. Since unmatched `{#`, `#}` tokens on separate lines would change from being text tokens to comment tokens. I expect that, as with unmatched `{%` tags, in the wild this would be pretty rare - but in my view, the unmatched comment token tags are a more dangerous problem. Unmatched `{%` will likely result in a `TemplateSyntaxError` in the new world, so devs will likely become aware of them, and be able to fix them pretty quickly. Comment token tags on different lines however, will simply become comments and will fail silently.

The same could be said about rogue `{{` tokens. All of this is a bit of a long-winded way of saying, we should minimise breaking changes as far as we can, unless there is a compelling reason to include them - even if such breaking changes are likely very rare.

Reply:
I've thought about this and while I agree that it's unfortunate that this hits comments and variables too, this is not really avoidable I think without making the regex much more complex. We don't want multiple regexes, and as you say, it's a big deal to change from a simple regex implementation to a custom parser.

Reply:
For clarity, the comment I made about moving away from a regex-based implementation is unrelated to my thoughts above about comment tokens and variable tokens - I'm very much still in two minds about a more radical shift away from a regex-based lexer. 

If we simply want to adapt the current lexer to introduce mutli-line block tokens, I don't think the regex will need to be too much more complicated. I've not tested the following, but I imagine something like this would do the trick:
```
tag_re = re.compile(r"({%(.|\n)*?%}|{{.*?}}|{#.*?#})")
```